rand_core/src/block.rs

```diff
@@ -352,27 +352,22 @@ where
 {
     #[inline]
     fn next_u32(&mut self) -> u32 {
-        let mut index = self.index * 2 - self.half_used as usize;
-        if index >= self.results.as_ref().len() * 2 {
+        if self.index >= self.results.as_ref().len() {
```

Comment:
This condition is different when `half_used` is true.

Note that for compatibility with `next_u64`, we use the half-entry at the *previous* index if `half_used`, thus the indexing of `self.results` below is also wrong.

So we still need a local: `let index = self.index - self.half_used as usize;`

Reply:
Fair enough! I'm a bit concerned that this was not caught by the tests.

Reply:
@dhardy So if I understand correctly, compatibility with `next_u64` means the following?

```rust
let rng1 = StdRng::from_entropy();
let rng2 = rng1.clone();
let (a, b, c) = (rng1.next_u32(), rng1.next_u64(), rng1.next_u32());
assert_eq!(rng2.next_u32(), a);
assert_eq!(rng2.next_u32(), c);
assert_eq!(rng2.next_u64(), b);
```

Reply:
@vks no, it's just that `next_u64` assumes that when `index < len`, then `results[index]` is a whole unused value. The method currently never *reads* `half_used`, though it does write. (That method can change if necessary.)

Reply:
But isn't that about accidentally reusing 32 bits that should be detectable in the example I gave above?